crates/ra_hir/src/code_model.rs

```diff
@@ -1362,6 +1390,74 @@ impl Type {
             ty: InEnvironment { value: ty, environment: self.ty.environment.clone() },
         }
     }
+
+    pub fn walk(&self, db: &dyn HirDatabase, mut cb: impl FnMut(Type)) {
+        // TypeWalk::walk does not preserve items order!
+        fn walk_substs(db: &dyn HirDatabase, substs: &Substs, cb: &mut impl FnMut(Type)) {
+            for ty in substs.iter() {
+                walk_ty(db, ty, cb);
+            }
+        }
+
+        fn walk_trait(
+            db: &dyn HirDatabase,
+            ty: Ty,
+            trait_ref: &TraitRef,
+            cb: &mut impl FnMut(Type),
+        ) {
+            let def_db: &dyn DefDatabase = db.upcast();
+            let resolver = trait_ref.trait_.resolver(def_db);
+            let krate = trait_ref.trait_.lookup(def_db).container.module(def_db).krate;
+            cb(Type::new_with_resolver_inner(db, krate, &resolver, ty));
+            walk_substs(db, &trait_ref.substs, cb);
+        }
+
+        fn walk_ty(db: &dyn HirDatabase, ty: &Ty, cb: &mut impl FnMut(Type)) {
+            let def_db: &dyn DefDatabase = db.upcast();
+            let ty = ty.strip_references();
+            match ty {
+                Ty::Apply(ApplicationTy { ctor, parameters }) => {
+                    match ctor {
+                        TypeCtor::Adt(adt) => {
+                            cb(Type::from_def(db, adt.module(def_db).krate, *adt));
```

Comment:
this isn't quite the right `Type` instance to construct here. Everywhere here, you should just use `derived` to construct the `Type`s; they are all still in the same environment as the original type.